### ISSUE TITLE
[Create Block] Adding a --convert-shortcode flag to scaffold a block from an existing shortcode

### ISSUE THREAD

**Issue by jonathanbossenger** [WordPress/gutenberg#43764]
## What problem does this address?
The inclusion of the --no-plugin flag got me thinking, what if we allowed developers to use Create Block to automatically scaffold code based on an existing shortcode?

Developers could run `npx @wordpress/create-block --convert-shortcode` inside of their plugin directory, and the tool could scan for any shortcodes in the plugin, and convert them to blocks.

## What is your proposed solution?
This is currently just an idea, and this ticket is being created so that we could explore that idea.

One suggestion could be to do the following

1. When using --convert-shortcode, require the developer to include the shortcode tag, callback function, and possible attributes.
2. Alternatively, require that the callback function have a valid PHP DocBlock, which includes the possible attributes, that way we only need to require the shortcode tag and callback function when using --covert-shortcode
3. Execute do_shortcode on the shortcode tag, passing a series of test variables as attributes, to generate the shortcode output
4. Using the shortcode tag, attributes, and HTML output, create the block code.
5. Optionally, include support for taking any CSS for the shortcode and migrating it over

This is one solution, which we could implement for the most straightforward of shortcode registrations, and then expand upon it as we discover more complex implementations.

**Comment by jonathanbossenger** [comment]
It could also be a requirement that the add_shortcode registration contains the PHP DocBlock, thereby only requiring that --convert-shortcode passed the shortcode tag.

**Comment by bph** [comment]
It's a great idea, however it would be great to have an accompanying blog series, that would tackle the meat of the matter of converting a shortcode to a block equivalent, with specific examples, approaches and a little theory as to why certain approaches might be best.

**Comment by jonathanbossenger** [comment]
That's a great idea and one that already has a series of tutorials Learn WordPress:

- [Part 1 - Using Create Block Tool](https://learn.wordpress.org/tutorial/using-the-create-block-tool/)
- [Part 2 - Converting a Shortcode to a Block](https://learn.wordpress.org/tutorial/converting-a-shortcode-into-a-block/)
- [Part 3 - Styling the Block](https://learn.wordpress.org/tutorial/styling-your-wordpress-block/)
- [Part 4 (in review) Using Block Attributes to Enable User Editing](https://github.com/WordPress/Learn/issues/860)

Once part 4 is published, this could be turned into a blog post, with accompanying videos, perhaps for a core dev blog of some kind... 😁

**Comment by gziolo** [comment]
@jonathanbossenger, thank you for sharing you idea to add a new feature to Create Block.

> Developers could run npx @wordpress/create-block --convert-shortcode inside of their plugin directory, and the tool could scan for any shortcodes in the plugin, and convert them to blocks.

I have some questions that could help me understand the problem better. How viable would it be to scan the existing code and extract all the required data you need? How many plugins are there still that developers might want to convert to blocks? It's been already nearly 4 years since the block editor got introduced in WordPress core 😄 

> It's a great idea, however it would be great to have an accompanying blog series, that would tackle the meat of the matter of converting a shortcode to a block equivalent, with specific examples, approaches and a little theory as to why certain approaches might be best.

That feels like a better starting point if there is a growing demand for this type of work where shortcodes need to be converted to blocks.

**Comment by LukaszJaro** [comment]
This is cool, is there a beta available for this I can try out?

**Comment by jonathanbossenger** [comment]
> How many plugins are there still that developers might want to convert to blocks?

That's actually a very good, and probably the most important question. A quick search on [WPDirectory](https://wpdirectory.net/) shows that there are 64208 instances of add_shortcode and 7714 instances of register_block_type, which leads me to believe more than we realize. 

However, what I don't know is how many of these 64208 shortcodes need to be converted (ie outdated/unmaintained plugins). I'm also not aware of there are other ways to register blocks, that I might also need to search for. Finally, I don't know of a better way to determine these numbers.

I wonder if it would make sense to share this idea as a post on the make/core and make/plugins blogs on wordpress.org?

**Comment by alanef** [comment]
Just some thoughts.

**TL:DR**  Don't waste time on tools - create better 'how to' guides covering simple to complex

The WP TV tutorial above falls well short of what is required in 'converting  a shortcode to a block'  in that most dev would obviously keep the shortcode and having two sets of code one inside the block and one inside the shortcode is not desirable,  and additionally many shortcodes have multiple attributes which need to be implemented as block controls.

In reality ServerSide render https://developer.wordpress.org/block-editor/reference-guides/packages/packages-server-side-render/ is probably the only practical solution for complex shortcode  to give a good editing experience.

(I'm not expert but I have been, for a long while and nowhere near complete, trying to to convert an 80 attribute shortcode to a block and every step hit barriers that are complex to solve).

Whilst a tool to automate some of this may suit some developers - personally as a developer, I would prefer to understand the exact steps I'm taking so that I understand my code going forward,  so personally would like to have better 'how to' documentation from wp.org covering from the simplest to the most complex.  

My experience of trying to 'learn' from a point of zero knowledge over the last 9 months was at first thwarted by the lack of understandable  wp.org documentation  ( understandable to a PHP developer that is embarking on a voyage) AND the massive amount of out of date non wp.org tutorials on the web in general.

I would say, that the create block scaffold actually added to the early confusion especially as it does not take into consideration every plugin will have its unique arrangement of source code. It was only once I ignored that and did a Wp Learn with Ryan Welcher did things start to click. But then a long journey to actually move forward to anything production worthy in my own code.

**Comment by jonathanbossenger** [comment]
I'm going to self-close this issue, it would appear that this is not something the community wants or needs.
